core/sched.cc

```diff
@@ -295,8 +299,17 @@ void cpu::reschedule_from_interrupt(bool called_from_yield,
         } else if (!called_from_yield) {
             auto &t = *runqueue.begin();
             if (p->_realtime._priority > 0) {
-                // Only switch to a higher-priority realtime thread
-                if (t._realtime._priority <= p->_realtime._priority) {
+                // If the threads have the same realtime priority, then only reschedule
+                // if the currently executed thread has exceeded its time slice (if any).
+                if (t._realtime._priority == p->_realtime._priority &&
+                    ((!p->_realtime.has_slice() || p->_realtime.has_remaining()))) {
```

Comment:
> […] should stay running if `_time_slice` is 0 (run 'forever' until yields or waits) OR there is still time left to run per its `_time_slice`, right?

Yes `p->_realtime.has_slice()` checks if it has a time slice (i.e. `_time_slice` != 0), `p->_realtime.has_remaining()` checks if there is still time remaining on the slice (if it has one).

Note that, even if the thread has exceeded its time slice it may still be selected to run again if there is no thread with a higher priority. Hence, the priority comparison in the if condition.